src/lib/stack-client.js

```diff
@@ -0,0 +1,364 @@
+    resp => {
+      if (resp.status === 401) {
+        throw new UnauthorizedStackException()
+      } else if (resp.status === 204) {
+        window.location.reload()
+      }
+    }
+  ).catch(
+    () => { 
+      throw new UnavailableStackException()
+    }
+  )
+}
+
+/**
+ * Get a cozy URL object
+ *
+ * @function
+ * @returns {URL} 
+ */
+const getCozyURL = function() {
+  return new URL(getStackClient().uri)
+}
+
+/**
+ * Get a the cozy origin as an URL string
+ * 
+ * @function
+ * @returns {string}
+ */
+const getCozyURLOrigin = function() {
+  return getCozyURL().origin
+}
+
+/**
+ * @deprecated
+ * @private
+ */
+const updateAccessToken = function(token) {
+  throw new Error("updateAccessToken should not be used with a cozy-client instance initialization")
+}
+
+/**
+ * Fetch a resource with cozy-client
+ *
+ * Utility to maintain the compatibility with the legacy 
+ * standalone cozy-bar client
+ *
+ * @function
+ * @private
+ * @returns {Promise} the full raw JSON playload
+ */
+const fetchJSON = function(method, path, body, options={}) {
+  
+  // We mirror here a few lines from cozy-stack-client
+  // because we want a customized fetchJSON 
+  const headers = (options.headers = options.headers || {})
+  headers['Accept'] = 'application/json'
+  if (method !== 'GET' && method !== 'HEAD' && body !== undefined) {
+    if (!headers['Content-Type']) {
+      headers['Content-Type'] = 'application/json'
+      body = JSON.stringify(body)
+    }
+  }
+
+  return getStackClient().fetch(method, path, body, options).then(
+    resp => {
+      if (typeof errorStatuses[resp.status] === 'function') {
+        throw new errorStatuses[resp.status]()
+      }    
+      const contentType = resp.headers.get('content-type')
+      const isJson = contentType.includes('json')
+      if (!isJson) {
+        throw new Error("Server response not in JSON")
+      }
+      return resp.json()
+    }
+  )
+}
+
+/**
+ * Test if an error is from an HTTP 404
+ *
+ * @function
+ * @private
+ * @param {Function} error - received from a fetch
+ * @returns {boolean}
+ */
+const is404 = function(error) {
+  return ['NotFoundException', 'NotFound', 'FetchError'].includes(error.name) &&
+     error.status && 
+     error.status === 404
+} 
+
+/**
+ * Memoize the result of a function which does an HTTP fetch
+ * 
+ * If a call throws an error because the 
+ * underlying HTTP request returned a 404
+ * then this function returns a default value
+ * 
+ * In the absence of any other error, the result is
+ * cached and reused in the next call to the function.
+ * 
+ *
+ * @function
+ * @param {Function} fn - the function to memoize. It will be
+ *                        called without any parameter
+ * @param {Object} defaultValue - returned in case of 404
+ */                  
+const withCache = function(fn, defaultValue) {
+  let cache = undefined;
+  return function() {
+    if (cache === undefined) {
+      return fn().then(result => 
+        cache = result
```

Comment:
Isn't a `return` missing here? You save the result in the cache, but the result is no longer available to the called of `withCache`.

Reply:
And testing `withCache` would be a good idea. It's a simple function to test, but having the right implementation on the first try is harder.

Reply:
I used to not test internal non exported functions.

As long it's used only one time, let's consider it as an internal mechanism (in the legacy code it was merged in the `getContext` function, I only externalized it to be more readable).

If it ever begin to be used more in the future, then let's put it in its own file, with an `export` and a dedicated test